src/rules/ec2/EC2IMDSv2.ts

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+import { parse } from 'path';
+import { CfnResource, Stack } from 'aws-cdk-lib';
+import { CfnInstance, CfnLaunchTemplate } from 'aws-cdk-lib/aws-ec2';
+import { NagRuleCompliance, NagRules } from '../../nag-rules';
+
+// Check to see if the EC2 Istance is configured with a launch profile
+//Ec2imdsv2
+
+/**
+ * Determine if the EC2 Instance has a launch profile
+ * @param node the CfnResource to check
+ */
+export default Object.defineProperty(
+  (node: CfnResource): NagRuleCompliance => {
+    if (node instanceof CfnInstance) {
+      // to use imdsv2 there must be a launchtemplate.
+      if (node.launchTemplate == undefined) {
+        return NagRuleCompliance.NON_COMPLIANT;
+      }
+
+      // find the launchTemplate
+      let instanceLaunchTemplateName = Stack.of(node).resolve(
+        node.launchTemplate
+      ).launchTemplateName;
+
+      // find the launchTemplate by name, and see if its got httpTokens set as 'required'
+      for (const child of Stack.of(node).node.findAll()) {
+        if (child instanceof CfnLaunchTemplate) {
+          if (
+            isMatchingLaunchTemplate(child, instanceLaunchTemplateName) &&
+            hasHttpTokens(child)
+          ) {
+            return NagRuleCompliance.COMPLIANT;
+          }
+        }
+      }
+
+      return NagRuleCompliance.NON_COMPLIANT;
+    } else {
+      return NagRuleCompliance.NOT_APPLICABLE;
+    }
+  },
+  'name',
+  { value: parse(__filename).name }
+);
+
+function isMatchingLaunchTemplate(
+  node: CfnLaunchTemplate,
+  launchTemplateName: string | undefined
+): boolean {
+  const templateName = NagRules.resolveResourceFromInstrinsic(
+    node,
+    node.launchTemplateName
```

Comment:
We have  to cover the case where a launch template could have been specifed by name or by Id..   

I'm not sure if we should be trapping the potential misconfiguration in the test. ( if both Id and name have been specified , this might get picked up by the cdk construct for instance, but i'm not sure.


```function isMatchingLaunchTemplate(
  node: CfnLaunchTemplate,
  launchTemplateName?: string | undefined,
  launchTemplateId?: string | undefined
): boolean {
  if (launchTemplateId !== undefined && launchTemplateName !== undefined) {
    // trap this, is an error, in configuration
    throw new Error('Should not have both a templateName and templateId');
  }

  var found: boolean = false;

  // test by templateName
  if (launchTemplateName) {
    const templateName = NagRules.resolveResourceFromInstrinsic(
      node,
      node.launchTemplateName
    );
    found = templateName === launchTemplateName;
  }

  if (launchTemplateId) {
    const templateId = NagRules.resolveResourceFromInstrinsic(node, node.ref);
    found = templateId === launchTemplateId;
  }
  return found;
}
```

Reply:
It's not the responsibility of cdk-nag to check for malformed CloudFormation.

We should just treat it as if both could be specified and flag as non compliant if either reference a template without imdsv2

Reply:
Ok, i'll remove that check.